app/src/main/java/org/fcitx/fcitx5/android/input/popup/PopupComponent.kt

```diff
@@ -222,12 +282,15 @@ class PopupComponent :
             when (this) {
                 is PopupAction.ChangeFocusAction -> outResult = changeFocus(viewId, x, y)
                 is PopupAction.DismissAction -> dismissPopup(viewId)
-                is PopupAction.PreviewAction -> showPopup(viewId, content, bounds)
+                is PopupAction.PreviewAction -> showPopup(viewId, content, bounds, style)
                 is PopupAction.PreviewUpdateAction -> updatePopup(viewId, content)
+                is PopupAction.PreviewArmedAction -> showingEntryUi[viewId]?.setArmed(armed)
                 is PopupAction.ShowKeyboardAction -> showKeyboard(viewId, keyboard, bounds)
                 is PopupAction.ShowMenuAction -> showMenu(viewId, menu, bounds)
                 is PopupAction.TriggerAction -> outAction = triggerFocused(viewId)
             }
         }
     }
 }
+
+private data class Quadruple<A, B, C, D>(val first: A, val second: B, val third: C, val fourth: D)
```

Comment:
This is the wrost data structure. You never know what's inside the Quadruple unless reading through the instance creating code. It's just bad.